docs/standard/native-interop/best-practices.md

```diff
@@ -251,6 +251,58 @@ finally
 }
 ```
 
+## Windows and non-Windows data type considerations
+
+There are types in the C language that have latitude in how they are defined. When writing cross-platform interop, cases can arise where platforms differ and can cause issues if not considered.
+
+### C `long`
+
+The `long` in C is defined to have ["at least 32"](https://en.cppreference.com/w/c/language/arithmetic_types) bits. This means there is a minimum required but platforms are not limited to that. The table below illustrates the differences in provided bits for the C `long` data type between platforms.
+
+|             | 32-bit | 64-bit |
+|:------------|:-------|:-------|
+| Windows     | 32     | 32     |
+| macOS/\*nix | 32     | 64     |
+
+These differences make authoring cross-platform P/Invokes difficult when the native function is defined to use `long` on all platforms. When targeting .NET 5 and prior .NET versions, you should declare separate Windows and non-Windows signatures to handle the problem.
+
+```csharp
+// Cross platform C function
+// void Function(long a);
+[DllImport("NativeLib", EntryPoint = "Function")]
+extern static void Function32(int a);
+
+#if WINDOWS
+    [DllImport("NativeLib", EntryPoint = "Function")]
+    extern static void Function64(int a);
+#else
+    [DllImport("NativeLib", EntryPoint = "Function")]
+    extern static void Function64(long a);
+#endif
```

Comment:
An alternative sample could be to use `int` on Windows and `IntPtr` on non-Windows and not split the code below on IntPtr.Size.

Reply:
That's shorter, but perhaps less clearer. Does this suggest we're missing an `nlong`/`nulong` ?

Reply:
That's what `CLong` and `CULong` are. This example is for before those types existed.

Reply:
>That's shorter, but IMO less clear.

I agree. This is a generalized pattern and people free to optimize it with that suggestion. I want to explicitly demonstrate the sizes. I think `IntPtr` is hiding too much here and is implying the `size_t` semantics here which this follows but that is incidental.

Reply:
I understand the desire to be explicit. Let's go with the example you made.